tests/mesh/mesh_function_dfem.C

```diff
@@ -0,0 +1,219 @@
+#include <cppunit/TestCase.h>
+#include <libmesh/restore_warnings.h>
+
+#include <libmesh/equation_systems.h>
+#include <libmesh/serial_mesh.h>
+#include <libmesh/mesh_generation.h>
+#include <libmesh/edge_edge2.h>
+#include <libmesh/face_quad4.h>
+#include <libmesh/face_tri3.h>
+#include <libmesh/cell_hex8.h>
+#include <libmesh/dof_map.h>
+#include <libmesh/linear_implicit_system.h>
+#include <libmesh/mesh_refinement.h>
+#include <libmesh/mesh_function.h>
+#include <libmesh/numeric_vector.h>
+
+#include "test_comm.h"
+
+using namespace libMesh;
+
+Number position_function (const Point& p,
+                          const Parameters&,
+                          const std::string&,
+                          const std::string&)
+{
+  if (p(1) > 0.0)
+    return 0.0;
+  else
+    return 1.0;
+}
+
+class MeshfunctionDFEM : public CppUnit::TestCase
+{
+  /**
+   * The goal of this test is to verify proper operation of mesh function
+   * for discontinuous shape functions and points close to the boundary
+   */
+public:
+  CPPUNIT_TEST_SUITE( MeshfunctionDFEM );
+
+  CPPUNIT_TEST( test_point_locator_dfem );
+
+  CPPUNIT_TEST( test_mesh_function_dfem );
+
+  CPPUNIT_TEST_SUITE_END();
+
+protected:
+  SerialMesh * _mesh;
+  PointLocatorBase * _point_locator;
+
+  void build_mesh()
+  {
+    _mesh = new SerialMesh(*TestCommWorld);
+
+    /*
+      (0,1)           (1,1)
+        x---------------x
+        |               |
+        |               |
+        |               |
+        |               |
+        |               |
+        x---------------x
+       (0,0)           (1,0)
+        |               |
+        |               |
+        |               |
+        |               |
+        x---------------x
+       (0,-1)          (1,-1)
+     */
+
+    _mesh->set_mesh_dimension(2);
+
+    _mesh->add_point( Point(0.0,-1.0), 4 );
+    _mesh->add_point( Point(1.0,-1.0), 5 );
+    _mesh->add_point( Point(1.0, 0.0), 1 );
+    _mesh->add_point( Point(1.0, 1.0), 2 );
+    _mesh->add_point( Point(0.0, 1.0), 3 );
+    _mesh->add_point( Point(0.0, 0.0), 0 );
+
+    {
+      Elem* elem_top = _mesh->add_elem( new Quad4 );
+      elem_top->set_node(0) = _mesh->node_ptr(0);
+      elem_top->set_node(1) = _mesh->node_ptr(1);
+      elem_top->set_node(2) = _mesh->node_ptr(2);
+      elem_top->set_node(3) = _mesh->node_ptr(3);
+
+      Elem* elem_bottom = _mesh->add_elem( new Quad4 );
+      elem_bottom->set_node(0) = _mesh->node_ptr(4);
+      elem_bottom->set_node(1) = _mesh->node_ptr(5);
+      elem_bottom->set_node(2) = _mesh->node_ptr(1);
+      elem_bottom->set_node(3) = _mesh->node_ptr(0);
+    }
+
+    // libMesh will renumber, but we numbered according to its scheme
+    // anyway. We do this because when we call uniformly_refine subsequenly,
+    // it's going use skip_renumber=false.
+    _mesh->prepare_for_use(false /*skip_renumber*/);
+
+    // get a point locator
+    _point_locator = _mesh->sub_point_locator().release();
```

Comment:
`release()` doesn't seem right, since I think it "steals" the `_point_locator` held internally by the Mesh.

Why not just let the Mesh manage the PointLocator lifetime, and get a `UniquePtr` to it when you use it in `test_point_locator_dfem()`?